plugin/src/main/java/org/opensearch/ml/action/models/DeleteModelTransportAction.java

```diff
@@ -44,36 +55,66 @@ public class DeleteModelTransportAction extends HandledTransportAction<ActionReq
-                    deleteModelChunks(modelId, deleteResponse, actionListener);
-                }
+            client.get(getRequest, ActionListener.runBefore(ActionListener.wrap(r -> {
+                if (r != null && r.isExists()) {
+                    try (XContentParser parser = createXContentParserFromRegistry(xContentRegistry, r.getSourceAsBytesRef())) {
+                        ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.nextToken(), parser);
+                        MLModel mlModel = MLModel.parse(parser);
+                        MLModelState mlModelState = mlModel.getModelState();
+                        if (mlModelState.equals(MLModelState.LOADED)
+                            || mlModelState.equals(MLModelState.LOADING)
+                            || mlModelState.equals(MLModelState.PARTIALLY_LOADED)) {
+                            actionListener
+                                .onFailure(
+                                    new Exception("Model cannot be deleted in loading or loaded state. Try unloading first and then delete")
+                                );
+                        } else {
+                            DeleteRequest deleteRequest = new DeleteRequest(ML_MODEL_INDEX, modelId);
+                            client.delete(deleteRequest, new ActionListener<DeleteResponse>() {
+                                @Override
+                                public void onResponse(DeleteResponse deleteResponse) {
+                                    deleteModelChunks(modelId, deleteResponse, actionListener);
+                                }
 
-                @Override
-                public void onFailure(Exception e) {
-                    log.error("Failed to delete model meta data for model: " + modelId, e);
-                    if (e instanceof ResourceNotFoundException) {
-                        deleteModelChunks(modelId, null, actionListener);
+                                @Override
+                                public void onFailure(Exception e) {
+                                    log.error("Failed to delete model meta data for model: " + modelId, e);
+                                    if (e instanceof ResourceNotFoundException) {
+                                        deleteModelChunks(modelId, null, actionListener);
+                                    }
+                                    actionListener.onFailure(e);
+                                }
+                            });
+                        }
+                    } catch (Exception e) {
+                        log.error("Failed to parse ml model" + r.getId(), e);
+                        actionListener.onFailure(e);
                     }
-                    actionListener.onFailure(e);
                 }
-            });
+            }, e -> { actionListener.onFailure(new MLResourceNotFoundException("Fail to find model")); }), () -> context.restore()));
```

Comment:
From line 81, we are going to restore context before running listener to delete model. Will it fail on security enabled cluster?

Reply:
Tested it on security enabled cluster and it works. Adding test results here.
<img width="1107" alt="Screenshot 2023-02-20 at 1 13 16 PM" src="https://user-images.githubusercontent.com/110800195/220201970-38505329-88bf-4099-b09f-658b091b59e2.png">